tutorials/batch-approve-swap.md

```diff
@@ -0,0 +1,419 @@
+
+For simplicity, we'll create a single script to handle all of the logic needed to swap DEV to DTOKs and back, called `swap.js`. We'll add this file to the `scripts` directory:
+
+```
+touch scripts/swap.js
+```
+
+### Create Contract Instances {: #create-contract-instances }
+
+We'll need to create contract instances for each of our contracts so that we can access each contract's functions. For this, we're going to use the `getContractAt` helper function of the Hardhat plugin.
+
+For this step, we're going to need the contract address of the `SimpleDex` contract. Then we'll be able to use the `SimpleDex` contract instance to retrieve the `DemoToken` contract address through the `token` function.
+
+We'll also need to add a contract instance for the batch precompile, which is located at `{{ networks.moonbase.precompiles.batch }}`.
+
+You can add the following code to the `swap.js` file:
+
+```js
+const simpleDexAddress = "INSERT-ADDRESS-OF-DEX";
+
+async function main() {
+  // Create instance of SimpleDex.sol
+  const simpleDex = await ethers.getContractAt(
+    "SimpleDex",
+    simpleDexAddress
+  );
+
+  // Create instance of DemoToken.sol
+  const demoTokenAddress = await simpleDex.token();
+  const demoToken = await ethers.getContractAt(
+    "DemoToken",
+    demoTokenAddress
+  );
+
+  // Create instance of Batch.sol
+  const batchAddress = "{{ networks.moonbase.precompiles.batch }}";
+  const batch = await ethers.getContractAt("Batch", batchAddress);
+}
+main();
+```
+
+### Add Check Balances Helper Function {: #add-function-to-check-balances }
+
+Next, we're going to create a helper function that will be used to check the balance of DTOK tokens the DEX and the signer account has. This will be particularly useful to see balance changes after the swaps are complete.
+
+Since the `DemoToken` contract has an ERC-20 interface, you can check the balance of DTOKs an account has using the `balanceOf` function. So, we'll call the `balanceOf` function, passing in the address of the signer and the DEX, and then print the formatted results in DTOKs to the terminal:
+
+```js
+async function checkBalances(demoToken) {
+  // Get the signer
+  const signer = (await ethers.getSigner()).address;
+
+  // Get the balance of the DEX and print it
+  const dexBalance = ethers.utils.formatEther(
+    await demoToken.balanceOf(simpleDexAddress)
+  );
+  console.log(`Dex ${simpleDexAddress} has a balance of: ${dexBalance} DTOKs`);
+
+  // Get the balance of the signer and print it
+  const signerBalance = ethers.utils.formatEther(
+    await demoToken.balanceOf(signer)
+  );
+  console.log(
+    `Account ${signer} has a balance of: ${signerBalance} DTOKs`
+  );
+}
+```
+
+### Swap DEV for Tokens {: #add-logic-to-swap-dev }
```

Comment:
My initial thought was to ax this section and just have the smart contract directly mint to the deployer's contract, since this section feels like setup that should be removed if possible so that the reader can get to the point of the tutorial faster.

```solidity
contract DemoToken is ERC20 {
    constructor(uint256 initialSupply) ERC20("DemoToken", "DTOK") {
        _mint(msg.sender, initialSupply / 2);
        _mint(tx.origin, initialSupply / 2);
    }
}
contract SimpleDex {
    // Made this payable so that liquidity can be added with deployment
    constructor() payable {
        // Mint 1000 DTOK tokens
        token = new DemoToken(1000000000000000000000);
    }
   //...
}
```

But it also looks like one of the main reasons that this step is needed is so that DEV liquidity exists for the contract. This can also be done in the constructor. If you choose not to remove this section, it would be good to also explain that liquidity is being added for the batch swap

Reply:
Yeah I agree that it's probably easiest to just do this in the constructor. Thanks for bringing this up!